### PR TITLE
rate limit in-flight ledger command submissions for trigger runners

### PR DESCRIPTION
**Pull request by carlpulley-da**
- [x] rate limit number of `SubmitRequest`s that ledger client will be allowed to submit
- [x] when number of in-flight command submissions is over a threshold value, fail any `SubmitRequest` as `UNAVAILABLE`

This PR partially resolves https://github.com/digital-asset/daml/pull/14453

ref: [gRPC StatusCode](https://grpc.github.io/grpc/cpp/namespacegrpc.html#aff1730578c90160528f6a8d67ef5c43b) for an explanation of gRPC status codes

Choice of `UNAVAILABLE` is based on the text:
> A litmus test that may help a service implementor in deciding between FAILED_PRECONDITION, ABORTED, and UNAVAILABLE: (a) Use UNAVAILABLE if the client can retry just the failing call. (b) Use ABORTED if the client should retry at a higher-level (e.g., restarting a read-modify-write sequence). (c) Use FAILED_PRECONDITION if the client should not retry until the system state has been explicitly fixed. E.g., if an "rmdir" fails because the directory is non-empty, FAILED_PRECONDITION should be returned since the client should not retry unless they have first fixed up the directory by deleting files from it. (d) Use FAILED_PRECONDITION if the client performs conditional REST Get/Update/Delete on a resource and the resource on the server does not match the condition. E.g., conflicting read-modify-write on the same resource.

CHANGELOG_BEGIN
* triggers may now control (via configuration) the rate at which they submit commands to the ledger. Command submission may be rate limited (to control trigger burst behaviour) and ledger submissions will *only* occur if in-flight command submissions are below threshold values.

CHANGELOG_END

### CODE REVIEW

File: triggers/runner/src/main/scala/com/digitalasset/daml/lf/engine/trigger/Runner.scala [hunk 1]
```diff
@@ -737,23 +746,62 @@ class Runner(
...
-      backoff = overloadedRetryDelay,
-      parallelism = maxParallelSubmissionsPerTrigger,
-      retryableSubmit,
-      submit,
-    )
-      .collect { case Some(err) => err }
+
+    val throttleFlow: Flow[SubmitRequest, SubmitRequest, NotUsed] =
+      Flow[SubmitRequest]
+        .throttle(maxSubmitRequests, maxSubmitRequestsDuration)
```
Comment (carlpulley-da) [inline triggers/runner/src/main/scala/com/digitalasset/daml/lf/engine/trigger/Runner.scala:760]: This flow (edge) will back pressure when `maxSubmitRequests` per `maxSubmitRequestsDuration` is exceeded

File: triggers/runner/src/main/scala/com/digitalasset/daml/lf/engine/trigger/Runner.scala [hunk 2]
```diff
@@ -737,23 +746,62 @@ class Runner(
...
+        .via(
+          retrying(
+            initialTries = maxTriesWhenOverloaded,
+            backoff = overloadedRetryDelay,
+            parallelism = maxParallelSubmissionsPerTrigger,
+            retryableSubmit,
+            submit,
+          )
+            .collect { case Some(err) => err }
+        )
```
Comment (carlpulley-da) [inline triggers/runner/src/main/scala/com/digitalasset/daml/lf/engine/trigger/Runner.scala:775]: This flow (edge) performs command submissions only when pending commands are below `maxInFlightCommands`

File: triggers/runner/src/main/scala/com/digitalasset/daml/lf/engine/trigger/Runner.scala [hunk 3]
```diff
@@ -737,23 +746,62 @@ class Runner(
...
+      Flow[SubmitRequest]
+        .filter { _ =>
+          pendingCommandIds.count(_._2 == SeenMsgs.Neither) > maxInFlightCommands
+        }
+        .map { request =>
+          SingleCommandFailure(
+            request.getCommands.commandId,
+            new StatusRuntimeException(UNAVAILABLE),
+          )
+        }
```
Comment (carlpulley-da) [inline triggers/runner/src/main/scala/com/digitalasset/daml/lf/engine/trigger/Runner.scala:786]: This flow (edge) simulates command submission failure (with an `UNAVAILABLE` gRPC status code) when pending commands are above `maxInFlightCommands`

File: triggers/runner/src/main/scala/com/digitalasset/daml/lf/engine/trigger/Runner.scala [hunk 4]
```diff
@@ -737,23 +746,62 @@ class Runner(
...
+      val throttle = gb.add(throttleFlow)
+      val submit = gb.add(submitFlow)
+      val failure = gb.add(failureFlow)
+
+      // format: off
+      throttle ~> broadcast ~> submit  ~> merge
+                  broadcast ~> failure ~> merge
+      // format: on
+
+      FlowShape(throttle.in, merge.out)
```
Comment (carlpulley-da) [inline triggers/runner/src/main/scala/com/digitalasset/daml/lf/engine/trigger/Runner.scala:801]: Resultant flow uses the `throttleFlow` edge in port as input and the `merge` node out port as output.

File: triggers/runner/src/main/scala/com/digitalasset/daml/lf/engine/trigger/Runner.scala [hunk 5]
```diff
@@ -737,23 +746,62 @@ class Runner(
...
+        .filter { _ =>
+          pendingCommandIds.count(_._2 == SeenMsgs.Neither) > maxInFlightCommands
+        }
+        .map { request =>
+          SingleCommandFailure(
+            request.getCommands.commandId,
+            new StatusRuntimeException(UNAVAILABLE),
+          )
+        }
+    val graph = GraphDSL.create() { implicit gb =>
```
Comment (carlpulley-da) [inline triggers/runner/src/main/scala/com/digitalasset/daml/lf/engine/trigger/Runner.scala:787]: As there is no way to merge the out ports of 2 flows, we are forced to use the graph DSL - sorry everyone!

File: triggers/runner/src/main/scala/com/digitalasset/daml/lf/engine/trigger/Runner.scala [hunk 6]
```diff
@@ -790,6 +838,15 @@ object Runner extends StrictLogging {
   val maxParallelSubmissionsPerTrigger = 8
   val maxTriesWhenOverloaded = 6
 
+  /** Maximum number of in-flight commands that we shall allow *before* the ledger
+    *  client automatically fails submit requests
+    */
+  val maxInFlightCommands = 42
```
Comment (carlpulley-da) [inline triggers/runner/src/main/scala/com/digitalasset/daml/lf/engine/trigger/Runner.scala:844]: Does anyone have any sensible values for this (its currently guessed!)?

File: triggers/runner/src/main/scala/com/digitalasset/daml/lf/engine/trigger/Runner.scala [hunk 7]
```diff
@@ -737,23 +745,63 @@ class Runner(
...
-    )
-      .collect { case Some(err) => err }
+
+    val throttleFlow: Flow[SubmitRequest, SubmitRequest, NotUsed] =
+      Flow[SubmitRequest]
+        .throttle(maxSubmitRequests, maxSubmitRequestsDuration)
+    val submitFlow: Flow[SubmitRequest, SingleCommandFailure, NotUsed] =
+      Flow[SubmitRequest]
+        .filter { _ =>
+          pendingCommandIds.count(_._2 == SeenMsgs.Neither) <= maxInFlightCommands
```
Comment (carlpulley-da) [inline triggers/runner/src/main/scala/com/digitalasset/daml/lf/engine/trigger/Runner.scala:763]: Should we be concerned about the use of `count` here? Maybe we should maintain some in-flight count as we maintain `pendingCommandIds`?

File: triggers/runner/src/main/scala/com/digitalasset/daml/lf/engine/trigger/Runner.scala [hunk 8]
```diff
@@ -737,23 +745,63 @@ class Runner(
...
+        }
+        .via(
+          retrying(
```
Comment (carlpulley-da) [inline triggers/runner/src/main/scala/com/digitalasset/daml/lf/engine/trigger/Runner.scala:766]: Using [groupedWithin](https://doc.akka.io/api/akka/2.6/akka/stream/scaladsl/Flow.html#groupedWithin(n:Int,d:scala.concurrent.duration.FiniteDuration):FlowOps.this.Repr[Seq[Out]]), we should be able to batch send ledger client `SubmitRequest`s. Is there any desire to do this on a separate PR?

File: triggers/runner/src/main/scala/com/digitalasset/daml/lf/engine/trigger/Runner.scala [hunk 9]
```diff
@@ -790,6 +838,15 @@ object Runner extends StrictLogging {
   val maxParallelSubmissionsPerTrigger = 8
   val maxTriesWhenOverloaded = 6
 
+  /** Maximum number of in-flight commands that we shall allow *before* the ledger
+    *  client automatically fails submit requests
+    */
+  val maxInFlightCommands = 42
+
+  /** Used to control rate at which we throttle ledger client submission requests */
+  val maxSubmitRequests = 100
```
Comment (remyhaemmerle-da) [inline triggers/runner/src/main/scala/com/digitalasset/daml/lf/engine/trigger/Runner.scala:847]: Community edition back pressure at `100` I will go with 20 or max 50.
```suggestion
  val maxSubmitRequests = 50
```

Anyway this should be configurable. (other PR).

File: triggers/runner/src/main/scala/com/digitalasset/daml/lf/engine/trigger/Runner.scala (hunk 6 above)
Comment (remyhaemmerle-da) [reply triggers/runner/src/main/scala/com/digitalasset/daml/lf/engine/trigger/Runner.scala:844]: The community edition start to back pressure at 100. 
so between 20 and 50 max.

Anyway this should be configurable.

File: triggers/runner/src/main/scala/com/digitalasset/daml/lf/engine/trigger/Runner.scala [hunk 10]
```diff
@@ -737,23 +746,62 @@ class Runner(
...
+
+      val broadcast = gb.add(Broadcast[SubmitRequest](2))
+      val merge = gb.add(Merge[SingleCommandFailure](2))
+      val throttle = gb.add(throttleFlow)
+      val submit = gb.add(submitFlow)
+      val failure = gb.add(failureFlow)
+
+      // format: off
+      throttle ~> broadcast ~> submit  ~> merge
+                  broadcast ~> failure ~> merge
```
[the remark this answers is not in the document]
Comment (remyhaemmerle-da) [reply triggers/runner/src/main/scala/com/digitalasset/daml/lf/engine/trigger/Runner.scala:798]: :cry:

File: triggers/runner/src/main/scala/com/digitalasset/daml/lf/engine/trigger/Runner.scala [hunk 11]
```diff
@@ -790,6 +838,15 @@ object Runner extends StrictLogging {
...
   val maxTriesWhenOverloaded = 6
 
+  /** Maximum number of in-flight commands that we shall allow *before* the ledger
+    *  client automatically fails submit requests
+    */
+  val maxInFlightCommands = 42
+
+  /** Used to control rate at which we throttle ledger client submission requests */
+  val maxSubmitRequests = 100
+  val maxSubmitRequestsDuration: FiniteDuration = 1.second
```
[the remark this answers is not in the document]
Comment (remyhaemmerle-da) [reply triggers/runner/src/main/scala/com/digitalasset/daml/lf/engine/trigger/Runner.scala:848]: Depending on the ledger a submission need between 100ms to 5s.

**Review by akrmn — APPROVED** [review]
LGTM but I'd defer to others with more Scala experience
Thanks Carl!

File: triggers/runner/src/main/scala/com/digitalasset/daml/lf/engine/trigger/Runner.scala (hunk 6 above)
Comment (carlpulley-da) [reply triggers/runner/src/main/scala/com/digitalasset/daml/lf/engine/trigger/Runner.scala]: Added in a `trigger-config.conf` file to trigger resources. using `pureconfig` we can now configure the trigger CLI/service on launch using either:
- `java -Dtrigger.parallelism=2 ...`
- or, `TRIGGER_MAX_INFLIGHT_COMMANDS=100; java ...`

This should also enable easier testing (in their) of any throttling or in-flight rate limiting behaviour that we have.

File: triggers/runner/src/main/scala/com/digitalasset/daml/lf/engine/trigger/Runner.scala [hunk 12]
```diff
@@ -782,13 +820,17 @@ class Runner(
...
+
+  private final case class RunnerConfig(
+      parallelism: Int,
+      maxRetries: Int,
+      maxInflightCommands: Int,
+      submission: RunnerSubmissionConfig,
+  )
+
+  private val config: RunnerConfig =
+    ConfigSource.resources("trigger-config.conf").at("trigger").loadOrThrow[RunnerConfig]
```
Comment (carlpulley-da) [inline triggers/runner/src/main/scala/com/digitalasset/daml/lf/engine/trigger/Runner.scala:833]: Loads `trigger` config namespace into the `RunnerConfig` case class. This loads once when the `Runner` companion object is loaded by the JVM. Any config load failure throws and so should cause the CLI/service to crash.

Comment (carlpulley-da) [reply triggers/runner/src/main/scala/com/digitalasset/daml/lf/engine/trigger/Runner.scala:833]: With the case class `'Int` types, we could introduce refined types to further validate (at config load time) that these values were (say) > 0.

**Review by remyhaemmerle-da — APPROVED** [review]
LGTM. Thanks

File: triggers/runner/src/main/scala/com/digitalasset/daml/lf/engine/trigger/Runner.scala (hunk 6 above)
Comment (carlpulley-da) [reply triggers/runner/src/main/scala/com/digitalasset/daml/lf/engine/trigger/Runner.scala]: As discussed offline, I'll remove this config code for now and add it into a separate PR

Comment (carlpulley-da) [reply triggers/runner/src/main/scala/com/digitalasset/daml/lf/engine/trigger/Runner.scala]: Set this value to be 50 (the upper value in the range suggested).

File: triggers/runner/src/main/scala/com/digitalasset/daml/lf/engine/trigger/Runner.scala (hunk 12 above)
Comment (carlpulley-da) [reply triggers/runner/src/main/scala/com/digitalasset/daml/lf/engine/trigger/Runner.scala]: These changes will be moved to a separate PR

File: triggers/runner/src/main/scala/com/digitalasset/daml/lf/engine/trigger/Runner.scala (hunk 6 above)
Comment (carlpulley-da) [reply triggers/runner/src/main/scala/com/digitalasset/daml/lf/engine/trigger/Runner.scala]: PR for trigger runner configuration is https://github.com/digital-asset/daml/pull/15482

File: triggers/runner/src/main/scala/com/digitalasset/daml/lf/engine/trigger/Runner.scala (hunk 12 above)
Comment (carlpulley-da) [reply triggers/runner/src/main/scala/com/digitalasset/daml/lf/engine/trigger/Runner.scala]: PR for trigger runner configuration is https://github.com/digital-asset/daml/pull/15482
